{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import sys, collections, os, argparse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build Index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Something is messed up in this workflow where we end up with MUSG instead of MUST for the ensembl \n",
    "# Right now I am fixing this in the other notebook, so it should be okay in the end, but just a note if I work on this \n",
    "# in the future"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mouse transcriptome already downloaded!\n"
     ]
    }
   ],
   "source": [
    "#Check if the file was downloaded already before doing curl:\n",
    "if not (os.path.isfile('Mus_musculus.GRCm38.cdna.all.fa.gz')): \n",
    "    # the `!` means we're running a command line statement (rather than python)  \n",
    "    !curl -O ftp://ftp.ensembl.org/pub/release-94/fasta/mus_musculus/cdna/Mus_musculus.GRCm38.cdna.all.fa.gz\n",
    "else: print('Mouse transcriptome already downloaded!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mouse transcript index already exist!\n"
     ]
    }
   ],
   "source": [
    "### Now we can build the index\n",
    "if not (os.path.isfile('mouse_transcriptome_index.idx')): \n",
    "    !kallisto index -i mouse_transcriptome_index.idx Mus_musculus.GRCm38.cdna.all.fa.gz\n",
    "else: print ('Mouse transcript index already exist!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download human and mouse reference transcriptomes from ensembl\n",
    "In order to do that we first download the human and mouse transcriptomes from ensembl. You can see the reference genomes they have at https://uswest.ensembl.org/info/data/ftp/index.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 28.0M  100 28.0M    0     0  10.1M      0  0:00:02  0:00:02 --:--:-- 10.1M\n"
     ]
    }
   ],
   "source": [
    "#Check if the file was downloaded already before doing curl:\n",
    "if not (os.path.isfile('Mus_musculus.GRCm38.94.gtf.gz')) or os.path.isfile('Mus_musculus.GRCm38.94.gtf'): \n",
    "    # the `!` means we're running a command line statement (rather than python)  \n",
    "    !curl -O ftp://ftp.ensembl.org/pub/release-94/gtf/mus_musculus/Mus_musculus.GRCm38.94.gtf.gz\n",
    "else: print('Mouse GTF already downloaded!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./Mus_musculus.GRCm38.94.gtf.gz:\t 96.2% -- replaced with ./Mus_musculus.GRCm38.94.gtf\n"
     ]
    }
   ],
   "source": [
    "# Unzip the files\n",
    "!gunzip -v -f ./Mus_musculus.GRCm38.94.gtf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create transcript_to_gene.tsv\n",
    "Now we can use the cells below to parse the GTF file and keep only the transcript mapping as a tsv file in the format below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_transcript_list(input, use_name = True, use_version = True):\n",
    "    r = {}\n",
    "    for line in input:\n",
    "        if len(line) == 0 or line[0] == '#':\n",
    "            continue\n",
    "        l = line.strip().split('\\t')\n",
    "        if l[2] == 'transcript':\n",
    "            info = l[8]\n",
    "            d = {}\n",
    "            for x in info.split('; '):\n",
    "                x = x.strip()\n",
    "                p = x.find(' ')\n",
    "                if p == -1:\n",
    "                    continue\n",
    "                k = x[:p]\n",
    "                p = x.find('\"',p)\n",
    "                p2 = x.find('\"',p+1)\n",
    "                v = x[p+1:p2]\n",
    "                d[k] = v\n",
    "\n",
    "\n",
    "            if 'transcript_id' not in d or 'gene_id' not in d:\n",
    "                continue\n",
    "\n",
    "            tid = d['transcript_id']\n",
    "            gid = d['gene_id']\n",
    "            if use_version:\n",
    "                if 'transcript_version' not in d or 'gene_version' not in d:\n",
    "                    continue\n",
    "\n",
    "                tid += '.' + d['transcript_version']\n",
    "                gid += '.' + d['gene_version']\n",
    "            gname = None\n",
    "            if use_name:\n",
    "                if 'gene_name' not in d:\n",
    "                    continue\n",
    "                gname = d['gene_name']\n",
    "\n",
    "            if tid in r:\n",
    "                continue\n",
    "\n",
    "            r[tid] = (gid, gname)\n",
    "    return r\n",
    "\n",
    "\n",
    "\n",
    "def print_output(output, r, use_name = True):\n",
    "    for tid in r:\n",
    "        if use_name:\n",
    "            output.write(\"%s\\t%s\\t%s\\n\"%(tid, r[tid][0], r[tid][1]))\n",
    "        else:\n",
    "            output.write(\"%s\\t%s\\n\"%(tid, r[tid][0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created mouse_transcript_to_gene.tsv file\n"
     ]
    }
   ],
   "source": [
    "with open('./Mus_musculus.GRCm38.94.gtf') as file:\n",
    "    r = create_transcript_list(file, use_name = True, use_version = True)\n",
    "with open('mouse_transcript_to_gene.tsv', \"w+\") as output:\n",
    "    print_output(output, r, use_name = True)\n",
    "print('Created mouse_transcript_to_gene.tsv file')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[index] k-mer length: 31\n",
      "[index] number of targets: 115,270\n",
      "[index] number of k-mers: 98,989,067\n",
      "[index] number of equivalence classes: 419,171\n",
      "[quant] will process sample 1: 200630Van/D20-2488-4537L/200630Van_D20-2488_1_sequence.fastq\n",
      "                               200630Van/D20-2488-4537L/200630Van_D20-2488_2_sequence.fastq\n",
      "[quant] will process sample 2: 200630Van/D20-3269-4585G/200630Van_D20-3269_1_sequence.fastq\n",
      "                               200630Van/D20-3269-4585G/200630Van_D20-3269_2_sequence.fastq\n",
      "[quant] finding pseudoalignments for the reads ... done\n",
      "[quant] processed 856,330,008 reads, 596,570,566 reads pseudoaligned\n"
     ]
    }
   ],
   "source": [
    "!kallisto bus -i mouse_transcriptome_index.idx  -o out_gmp_merged -x 10xv2 -t 2 \\\n",
    "200630Van/D20-2488-4537L/200630Van_D20-2488_1_sequence.fastq 200630Van/D20-2488-4537L/200630Van_D20-2488_2_sequence.fastq \\\n",
    "200630Van/D20-3269-4585G/200630Van_D20-3269_1_sequence.fastq 200630Van/D20-3269-4585G/200630Van_D20-3269_2_sequence.fastq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load transcript to gene file\n",
    "tr2g = {}\n",
    "trlist = []\n",
    "with open('./mouse_transcript_to_gene.tsv') as f:\n",
    "    for line in f:\n",
    "        l = line.split()\n",
    "        tr2g[l[0]] = l[1]\n",
    "        trlist.append(l[0])\n",
    "\n",
    "genes = list(set(tr2g[t] for t in tr2g))\n",
    "\n",
    "# load equivalence classes\n",
    "ecs = {}\n",
    "with open('./out_gmp_merged/matrix.ec') as f:\n",
    "    for line in f:\n",
    "        l = line.split()\n",
    "        ec = int(l[0])\n",
    "        trs = [int(x) for x in l[1].split(',')]\n",
    "        ecs[ec] = trs\n",
    "        \n",
    "def ec2g(ec):\n",
    "    if ec in ecs:\n",
    "        return list(set(tr2g[trlist[t]] for t in ecs[ec]))        \n",
    "    else:\n",
    "        return []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Processing the BUS file\n",
    "For these notebooks we will work with the text file that BUStools produces, rather than the raw BUS file. To install BUStools see https://github.com/BUStools/bustools\n",
    "\n",
    "We discard any barcodes that don't have more 10 UMIs\n",
    "\n",
    "To produce the text file, starting with the output.bus file produced by kallisto, we first sort it on bustools:\n",
    "\n",
    "bustools sort -o output.sorted output.bus\n",
    "Then we convert it to txt:\n",
    "\n",
    "bustools text -o output.sorted.txt output.sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in 596570566 number of busrecords\r\n"
     ]
    }
   ],
   "source": [
    "#sort bus file\n",
    "!bustools sort -o ./out_gmp_merged/output_sorted.bus ./out_gmp_merged/output.bus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in 472774795 number of busrecords\r\n"
     ]
    }
   ],
   "source": [
    "!bustools text -o ./out_gmp_merged/output_sorted.txt ./out_gmp_merged/output_sorted.bus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading generated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load transcripts\n",
    "tr2g = {}\n",
    "trlist = []\n",
    "with open('./mouse_transcript_to_gene.tsv') as f:\n",
    "    for line in f:\n",
    "        l = line.split()\n",
    "        tr2g[l[0]] = l[1]\n",
    "        trlist.append(l[0])\n",
    "\n",
    "genes = list(set(tr2g[t] for t in tr2g))\n",
    "\n",
    "# load equivalence classes\n",
    "ecs = {}\n",
    "with open('./out_gmp_merged/matrix.ec') as f:\n",
    "    for line in f:\n",
    "        l = line.split()\n",
    "        ec = int(l[0])\n",
    "        trs = [int(x) for x in l[1].split(',')]\n",
    "        ecs[ec] = trs\n",
    "        \n",
    "def ec2g(ec):\n",
    "    if ec in ecs:\n",
    "        return list(set(tr2g[trlist[t]] for t in ecs[ec]))        \n",
    "    else:\n",
    "        return []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "5000000\n",
      "10000000\n",
      "15000000\n",
      "20000000\n",
      "25000000\n",
      "30000000\n",
      "35000000\n",
      "40000000\n",
      "45000000\n",
      "50000000\n",
      "55000000\n",
      "60000000\n",
      "65000000\n",
      "70000000\n",
      "75000000\n",
      "80000000\n",
      "85000000\n",
      "90000000\n",
      "95000000\n",
      "100000000\n",
      "105000000\n",
      "110000000\n",
      "115000000\n",
      "120000000\n",
      "125000000\n",
      "130000000\n",
      "135000000\n",
      "140000000\n",
      "145000000\n",
      "150000000\n",
      "155000000\n",
      "160000000\n",
      "165000000\n",
      "170000000\n",
      "175000000\n",
      "180000000\n",
      "185000000\n",
      "190000000\n",
      "195000000\n",
      "200000000\n",
      "205000000\n",
      "210000000\n",
      "215000000\n",
      "220000000\n",
      "225000000\n",
      "230000000\n",
      "235000000\n",
      "240000000\n",
      "245000000\n",
      "250000000\n",
      "255000000\n",
      "260000000\n",
      "265000000\n",
      "270000000\n",
      "275000000\n",
      "280000000\n",
      "285000000\n",
      "290000000\n",
      "295000000\n",
      "300000000\n",
      "305000000\n",
      "310000000\n",
      "315000000\n",
      "320000000\n",
      "325000000\n",
      "330000000\n",
      "335000000\n",
      "340000000\n",
      "345000000\n",
      "350000000\n",
      "355000000\n",
      "360000000\n",
      "365000000\n",
      "370000000\n",
      "375000000\n",
      "380000000\n",
      "385000000\n",
      "390000000\n",
      "395000000\n",
      "400000000\n",
      "405000000\n",
      "410000000\n",
      "415000000\n",
      "420000000\n",
      "425000000\n",
      "430000000\n",
      "435000000\n",
      "440000000\n",
      "445000000\n",
      "450000000\n",
      "455000000\n",
      "460000000\n",
      "465000000\n",
      "470000000\n"
     ]
    }
   ],
   "source": [
    "# load kallisto bus output dataset\n",
    "cell_gene = collections.defaultdict(lambda: collections.defaultdict(float))\n",
    "pbar=None\n",
    "pumi=None\n",
    "with open('./out_gmp_merged/output_sorted.txt') as f:\n",
    "    gs = set()\n",
    "    c = 0\n",
    "    for line in f:\n",
    "        if (c % 5000000) == 0:\n",
    "            print(c)\n",
    "        l = line.split()\n",
    "        barcode,umi,ec,count = line.split()\n",
    "        ec = int(ec)\n",
    "        \n",
    "        if barcode == pbar:\n",
    "            # same barcode\n",
    "            if umi == pumi:\n",
    "                # same UMI, let's update with intersection of genelist\n",
    "                gl = ec2g(ec)\n",
    "                gs.intersection_update(gl)\n",
    "            else:\n",
    "                # new UMI, process the previous gene set\n",
    "                for g in gs:\n",
    "                    cell_gene[barcode][g] += 1.0/len(gs)\n",
    "                # record new umi, reset gene set\n",
    "                pumi = umi\n",
    "                gs = set(ec2g(ec))\n",
    "        else:\n",
    "            # work with previous gene list\n",
    "            for g in gs:\n",
    "                cell_gene[pbar][g] += 1.0/len(gs)\n",
    "            \n",
    "            if sum(cell_gene[pbar][g] for g in cell_gene[pbar]) < 10:\n",
    "                del cell_gene[pbar]\n",
    "            \n",
    "            pbar = barcode\n",
    "            pumi = umi\n",
    "            \n",
    "            gs = set(ec2g(ec))\n",
    "        c += 1\n",
    "    #remember the last gene\n",
    "    for g in gs:\n",
    "        cell_gene[pbar][g] += 1.0/len(gs)\n",
    "        \n",
    "    if sum(cell_gene[pbar][g] for g in cell_gene[pbar]) < 10:\n",
    "        del cell_gene[pbar]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_hist = collections.defaultdict(int)\n",
    "for barcode in cell_gene:\n",
    "    cg = cell_gene[barcode]\n",
    "    s = len([cg[g] for g in cg])\n",
    "    barcode_hist[barcode] += s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "385190\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEGCAYAAACKB4k+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAez0lEQVR4nO3dfZRcVZnv8e+P8CIS3sEeDGgHgtwbyXXGtGLUqx2VS1CCykWHDMsRROMbvozoGC6O4nI56mjuUl4czCiisyI9ER2BEIdRpFGvLIjBwRAxGDGOHZDAxGkNixFjnvvH2V2pNFXVp6tr10v377NWrT5nn7enTnfV03vvc/ZRRGBmZgawT6cDMDOz7uGkYGZmFU4KZmZW4aRgZmYVTgpmZlaxb6cDmIqjjjoq+vv7m9r20Ucf5aCDDmptQC3mGFvDMbaGY2yNbohxw4YNj0TE0TUXRkTPvYClwKp58+ZFs2699damt20Xx9gajrE1HGNrdEOMwA+jzvdrTzYfRcSNEbH80EMP7XQoZmbTSk8mBTMzy8NJwczMKnoyKUhaKmnV6Ohop0MxM5tWejIpuE/BzCyPnkwKZmaWh5OCmZlV9GRScJ+CmVkePZkUWtGnsHHbKP0rbqJ/xU0tjMzMrLf1ZFIwM7M8nBTMzKzCScHMzCqcFMzMrMJJwczMKnoyKfiSVDOzPHoyKXiYCzOzPHoyKZiZWR5OCmZmVuGkYGZmFU4KZmZW4aRgZmYVXZMUJA1K+p6kqyQNdjoeM7OZKGtSkHS1pO2S7hlXvkTSZklbJK1IxQHsBJ4EjOSMy8zMastdU7gGWFJdIGkWcCVwOjAfWCZpPvC9iDgdeD/w4cxxmZlZDVmTQkR8F9gxrvi5wJaIuD8iHgeGgFdGxO60/DfAATnjMjOz2hQReQ8g9QNrI+LkNH82sCQi3pjmXwecAnwHOA04DPj7iBius7/lwHKAvr6+hUNDQ03FtX3HKA89VkwvmNOdd0bv3LmT2bNndzqMhhxjazjG1nCM5SxevHhDRAzUWrZvu4MBVKMsIuLrwNcn2jgiVkl6EFh68MEHLxwcHGwqiMtXX8/KjcXb33puc/vIbXh4mGbfX7s4xtZwjK3hGKeuE1cfjQDHVc0fCzwwmR147CMzszw6kRTWAydKmitpf+Ac4IbJ7MCjpJqZ5ZH7ktRrgduBkySNSLogInYBFwI3A/cCayJi02T265qCmVkeWfsUImJZnfJ1wLpm9ytpKbB03rx5ze7CzMxq6Jo7mifDNQUzszx6MimYmVkePZkU3NFsZpZHTyYFNx+ZmeXRk0nBzMzy6Mmk4OYjM7M8ejIpuPnIzCyPnkwKZmaWR08mBTcfmZnl0ZNJwc1HZmZ59GRSMDOzPJwUzMyswknBzMwqejIpuKPZzCyPnkwK7mg2M8ujE89o7jr9K26qTG/9+Cs6GImZWWf1ZE3BzMzycE2hA1wzMbNu5ZqCmZlV9GRS8NVHZmZ59GRS8NVHZmZ59GRSMDOzPNzR3CbVnctmZt3KNQUzM6twUjAzswonBTMzq3BSMDOzikklBUn7SDokVzCSDpK0QdIZuY5hZmb1TZgUJH1F0iGSDgJ+AmyW9L4yO5d0taTtku4ZV75E0mZJWyStqFr0fmDNZN6AmZm1TpmawvyI+C3wKmAd8DTgdSX3fw2wpLpA0izgSuB0YD6wTNJ8SS+jSDoPldy3mZm1mCKi8QrSJuBPga8AV0TEbZLujohnlTqA1A+sjYiT0/wi4NKIOC3NX5xWnQ0cRJEoHgNeHRG7a+xvObAcoK+vb+HQ0FCZMJ5g+45RHnrsieUL5uS5S3rjttpDcjQ63s6dO5k9e3aWeFrFMbaGY2wNx1jO4sWLN0TEQK1lZW5e+xywFbgb+K6kpwO/nUI8c4BfVc2PAKdExIUAks4DHqmVEAAiYhWwCmBgYCAGBwebCuLy1dezcuMT3/7Wc5vb30TOq3PzWqPjDQ8P0+z7axfH2BqOsTUc49RNmBQi4jLgsqqiX0paPIVjqtZhqo53zYQ7kJYCS+fNmzeFMMzMbLwJk4KkPuBvgadGxOmS5gOLgC80ecwR4Liq+WOBB5rcV8/zsxXMrJuU6Wi+BrgZeGqavw949xSOuR44UdJcSfsD5wA3TGYHHiXVzCyPMknhqIhYA+wGiIhdwB/L7FzStcDtwEmSRiRdkLa/kCLR3AusiYhNTUVvZmYtVaaj+VFJR5La/SU9Dyj1dJuIWFanfB3F5a1NcZ+CmVkeZWoK76Fo3jlB0v8Dvgy8I2tUE3DzkZlZHmWuPrpL0ouBkyiuHNocEX/IHlkDrimYmeVRNylIOqvOomdIIiK+nimmCUXEjcCNAwMDb+pUDGZm01GjmsLS9PMpwPOB76T5xcAw0LGkYGZmedRNChFxPoCktRTjHz2Y5o+hGLuoY9x8ZGaWR5mO5v6xhJA8BDwjUzyluKPZzCyPMpekDku6GbiW4rLUc4Bbs0ZlZmYdUebqowslvRp4USpaFRH/nDesxtx8ZGaWR5maAsAPgF0UNYU784VTTs6rjzwWkZnNZGUGxHst8EmKK44EXC7pfRFxXebYel5/neGyzcy6VZmawiXAcyJiO4Cko4FvA04KZmbTTJmrj/YZSwjJf5TczszMekyZmsK/VF19BPDnwDfzhTQxdzSbmeUx4X/8EfE+ikdy/g/gWRRXH/117sAmiMn3KZiZZVCmo3kusG5srCNJB0rqj4ituYMzM7P2KtM38FXSA3aSP6YyMzObZsokhX0j4vGxmTS9f76QzMysU8okhYclnTk2I+mVwCP5QjIzs04pc/XRW4DVkq5I8yPA6/KFNDFffWRmlkfDmoKkfYCFEfE8YD7wzIh4fkT8vC3R1eGrj8zM8miYFCJiN3Bhmt4ZEb9rS1RmZtYRZfoUviXpvZKOk3TE2Ct7ZGZm1nZl+hTekH6+vaosgONbH46ZmXVSmecpzG1HIGZm1nmlnqcg6WSKjuYnjZVFxJdzBWVmZp1RZpiLDwGDFElhHXA68H3ASaHF/IAfM+u0Mh3NZwMvBX4dEedTDIp3QKsDkfTfJV0l6TpJb231/s3MbGJlksJj6dLUXZIOAbZTspNZ0tWStku6Z1z5EkmbJW2RtAIgIu6NiLcArwUGJvc2zMysFcr0KfxQ0mHAPwAbgJ2Uf07zNcAVVDU1SZoFXAmcSnF39HpJN0TET9JwGivSNh3n5hwzm2kUEeVXlvqBQyLix5PcZm1EnJzmFwGXRsRpaf5igIj4WNU2N0VEzW9hScuB5QB9fX0Lh4aGSsdfbfuOUR56rPz6C+ZM/u7pjdtGJ71N9fF27tzJ7Nmzm95HOzjG1nCMreEYy1m8ePGGiKjZIlP26qOzgBdS3J/wfaB0UqhhDvCrqvkR4BRJg8BZFP0V6+ptHBGrgFUAAwMDMTg42FQQl6++npUbS719ALaeO/njnFdV05isrecOMjw8TLPvr10cY2s4xtZwjFNX5uqjzwLz2PM4zjdLellEvL3BZg13WaMsImIYGC61Aw+IZ2aWRZl/lV8MnBypnUnSl4CNUzjmCHBc1fyxwAOT2UFE3AjcODAw8KYpxGFmZuOUSQqbgacBv0zzxzG15qP1wInpMZ/bgHOAv5jMDrq5ptA/hSYjM7NOq3tJqqQbJd0AHAncK2lY0q3AvcDRZXYu6VrgduAkSSOSLoiIXRQjr96c9rUmIjZNJmgPnW1mlkejmsKnprrziFhWp3wdDTqTJ9LNNQUzs15WNylExG3tDGQy3KdgZpZHmTuau46kpZJWjY42fx+AmZk9UU8mBfcpmOXRv+Kmymt8+cZto76QYgao23wk6ZaIeKmkT0TE+9sZlJm1z2S/6Out76FgpodGHc3HSHoxcKakIcbddBYRd2WNrAF3NJtNTZlE4GQxMzVKCh+kGJzuWOD/jlsWwEtyBTWRbutodpXaulU3DOrYDTFYeY2uProOuE7S30TER9oYU1fyH7b1Ov/zYmWUeUbzR9KQ1i9KRcMRsTZvWGY2Gf7Ct1YpMyDex4DnAqtT0bskvSAiLs4aWeOYOtqn4A+gWXMafXZcA+8OZS5JfQVwakRcHRFXA0tSWcf4klQzszzKPlDgMGBHmvY3sVkXcI3VciiTFD4G/CgNhieKvoWONR2Z2fTkizm6Q5mO5mslDQPPoUgK74+IX+cOrJFO9ymYdYprB5ZbqeajiHgQuCFzLKV1230KZtZarjV0TvmHFFtX8IdlZnCNYA//zbdXTw6INxN4ADKzJ6o3YJ+1TsOkIGkfSfe0KxgzM+ushs1HEbFb0t2SnhYR/96uoMzMJuJmpTzK9CkcA2ySdCfw6FhhRJyZLaoJ+Oojm47cJGLdoExS+HD2KCZppl19VO/Lwv8pmVmrlblP4TZJTwdOjIhvS3oyMCt/aGbTU60kf9GCXfhiQOsGZQbEexOwHDgCOAGYA1wFvDRvaNatqr/UrllyUAcj6Q1uFspv7BxftGAXg50NpeeV+dfk7RSjpN4BEBE/k/SUrFFZV3DzlPUi/91OTZmk8PuIeFwqnsYpaV+KJ6/ZDNLMf7sz7cM5096vTU9lksJtkv4PcKCkU4G3ATfmDcvaqVXNG+1oJum2L143DXW3bvt76QVlksIK4AJgI/BmYB3w+ZxBWX6t+jLbuG2U86awr9wf2v4VN3HRgl2ct+KmSe/fD6K3majM1Ue7JX2Jok8hgM0RkaX5SNKrKB7g8xTgyoj41xzHsfZrx5d/K9YpE5trBzadlbn66BUUVxv9nGLo7LmS3hwR3yxzAElXA2cA2yPi5KryJcBnKC5v/XxEfDwivgF8Q9LhwKcAJ4WS6n3p9tIXWC/FOpHp9F5sZinTfLQSWBwRWwAknQDcBJRKCsA1wBXAl8cKJM0CrgROBUaA9ZJuiIifpFU+kJbbNNRLfRg2fbh/oRxN1BIk6bsR8aKqeQG3VZdNeBCpH1g7VlOQtAi4NCJOS/NjT3L7eHp9KyK+XWdfyynum6Cvr2/h0NBQ2TD2sn3HKA891tSmbdN3IE3FuGDOniembtw22sKInqjZGNvJMbbGdIqx+jPSbjt37mT27NkdOz7A4sWLN0TEQK1ldWsKks5Kk5skrQPWUPQpvAZYP8WY5gC/qpofAU4B3gG8DDhU0ryIuGr8hhGxClgFMDAwEIODg00FcPnq61m5sbvvIL1owa7mYtz4aNVM3vfYdIxt5BhbYzrFuPXcwfzB1DE8PEyz31vt0OjsLa2afgh4cZp+GDh8isdVjbKIiMuAyybc2APimZllUTcpRMT5GY87AhxXNX8s8EDZjWfagHhmZu1S5uqjuRTNOv3V609x6Oz1wIlp39uAc4C/KLuxawpmNhW+B6W+Mg2E3wC+QHEX8+7JHkDStcAgcJSkEeBDEfEFSRcCN1Ncknp1RGwqu0/XFMzM8iiTFP4rtfU3JSKW1SlfR3F39KS5pmBmlkeZpPAZSR+iuJHs92OFEXFXtqgm4JqCmeXgexnKJYUFwOuAl7Cn+SjSfEe4pmBmlkeZpPBq4PiIeDx3MGW5pmBmlsc+Jda5GzgsdyBmZtZ5ZWoKfcBPJa1n7z6FqVySOiVuPjIzy6NMUvhQ9igmyc1HZmZ5lHmewm3tCMTMrJvM1BvcytzR/Dv2PJN5f2A/4NGIOCRnYGZm1n5lagoHV8+np6M9N1tEJbhPwcwsjzJXH+0lPR2tY/copBhujIjlhx7auTHRzcymozLNR2dVze4DDLCnOcnMzKaRMlcfVT9XYRewFXhllmjMzKyjyvQp5HyuQlPcp2Bmlkejx3F+sMF2EREfyRBPKb5Pwcwsj0Y1hUdrlB0EXAAcCXQsKZiZWR6NHse5cmxa0sHAu4DzgSFgZb3tzMyms+k+vHbDPgVJRwDvAc4FvgQ8OyJ+047AzMys/Rr1KXwSOAtYBSyIiJ1ti8rMzDqiUU3hIopRUT8AXCJprFwUHc0e5sLMZrTpOD5Soz6FSd/t3C6+JNXMLI+u/eJvxMNcmJnl0ZNJwczM8nBSMDOzCicFMzOrKDMgnpmZNanXbnZzTcHMzCq6JilIOl7SFyRd1+lYzMxmqqzNR5KuBs4AtkfEyVXlS4DPALOAz0fExyPifuACJwUz63X1bmrrBblrCtcAS6oLJM0CrgROB+YDyyTNzxyHmZmVkDUpRMR3gR3jip8LbImI+yPicYpRV/0kNzOzLqCIvI9bltQPrB1rPpJ0NrAkIt6Y5l8HnAJ8CPgocCpFk9LH6uxvObAcoK+vb+HQ0FBTcW3fMcpDjzW1adv0HYhjbAHH2BqOceoWzDmUnTt3Mnv27I7GsXjx4g0RMVBrWScuSVWNsoiI/wDeMtHGEbFK0oPA0oMPPnjh4OBgU0Fcvvp6Vm7s7ityL1qwyzG2gGNsDcc4dVvPHWR4eJhmv7faoRNXH40Ax1XNHws8MJkdeOwjM7M8OpEU1gMnSporaX/gHOCGyexA0lJJq0ZHR7MEaGY2U2VNCpKuBW4HTpI0IumCiNgFXAjcDNwLrImITZPZr2sKZmZ55L76aFlEHBMR+0XEsRHxhVS+LiKeEREnRMRHJ7tf1xTMrBf1r7iJjdtGu/o+hq65o3kyXFMwM8ujJ5OCmZnl0ZNJwc1HZmZ59GRScPORmVkePZkUzMwsj55MCm4+MjPLoyeTgpuPzMzy6MmkYGZmeXTvyFENSFoKLJ03b16nQzEza0q3Pru5J2sKbj4yM8ujJ5OCmZnl4aRgZmYVTgpmZlbhjmYzsy7W7g7pnqwpuKPZzCyPnkwKZmaWh5OCmZlVOCmYmVmFk4KZmVX46iMzsw7rpiEverKm4KuPzMzy6MmkYGZmeTgpmJlZhZOCmZlVOCmYmVmFk4KZmVU4KZiZWUXX3Kcg6SDgs8DjwHBErO5wSGZmM07WmoKkqyVtl3TPuPIlkjZL2iJpRSo+C7guIt4EnJkzLjMzqy1389E1wJLqAkmzgCuB04H5wDJJ84FjgV+l1f6YOS4zM6tBEZH3AFI/sDYiTk7zi4BLI+K0NH9xWnUE+E1ErJU0FBHn1NnfcmA5QF9f38KhoaGm4tq+Y5SHHmtq07bpOxDH2AKOsTUcY2tMFOOCOXuP1LBx22ip9SZj8eLFGyJioNayTvQpzGFPjQCKZHAKcBlwhaRXADfW2zgiVgGrAAYGBmJwcLCpIC5ffT0rN3ZNl0pNFy3Y5RhbwDG2hmNsjYli3Hru4F7z51WNi9RovVbpxNlTjbKIiEeB80vtwAPimZll0YlLUkeA46rmjwUe6EAcZmY2TieSwnrgRElzJe0PnAPcMJkdeJRUM7M8cl+Sei1wO3CSpBFJF0TELuBC4GbgXmBNRGzKGYeZmZWTtU8hIpbVKV8HrGt2v+5TMDPLoyeHuXDzkZlZHj2ZFCQtlbRqdLT29btmZtacnkwKrimYmeWR/Y7mnCQ9DPyyyc2PAh5pYTg5OMbWcIyt4RhboxtifHpEHF1rQU8nhamQ9MN6t3l3C8fYGo6xNRxja3R7jD3ZfGRmZnk4KZiZWcVMTgqrOh1ACY6xNRxjazjG1ujqGGdsn4KZmT3RTK4pmJnZOE4KZmZWMSOTQp1nRLfjuMdJulXSvZI2SXpXKj9C0rck/Sz9PDyVS9JlKc4fS3p21b5en9b/maTXZ4h1lqQfSVqb5udKuiMd75/SCLdIOiDNb0nL+6v2cXEq3yzptBbHd5ik6yT9NJ3PRd12HiX9Vfo93yPpWklP6obzWOvZ6a08d5IWStqYtrlMUq1nqEw2vk+m3/WPJf2zpMOqltU8P/U+5/V+B1ONsWrZeyWFpKPSfNvP4ZRExIx6AbOAnwPHA/sDdwPz23TsY4Bnp+mDgfsonlP9d8CKVL4C+ESafjnwTYoHEz0PuCOVHwHcn34enqYPb3Gs7wG+QvEoVYA1wDlp+irgrWn6bcBVafoc4J/S9Px0bg8A5qZzPquF8X0JeGOa3h84rJvOI8UTBn8BHFh1/s7rhvMIvAh4NnBPVVnLzh1wJ7AobfNN4PQWxPe/gH3T9Ceq4qt5fmjwOa/3O5hqjKn8OIoRoH8JHNWpczilv492HahbXulE31w1fzFwcYdiuR44FdgMHJPKjgE2p+nPAcuq1t+cli8DPldVvtd6LYjrWOAW4CXA2vSH+UjVh7JyDtMHYFGa3jetp/HntXq9FsR3CMUXrsaVd815ZM9jZ49I52UtcFq3nEegn72/dFty7tKyn1aV77Ves/GNW/ZqYHWarnl+qPM5b/S33IoYgeuAZwFb2ZMUOnIOm33NxOajWs+IntPuIFLzwJ8BdwB9EfEgQPr5lLRavVhzv4dPA38N7E7zRwL/GcWzMMYfrxJLWj6a1s8Z4/HAw8AXVTRxfV7SQXTReYyIbcCngH8HHqQ4LxvorvNYrVXnbk6azhnvGyj+e24mvkZ/y1Mi6UxgW0TcPW5RN57DumZiUqj5jOi2BiDNBr4GvDsiftto1Rpl0aC8FbGdAWyPiA0l4mi0LOd53pei6v73EfFnwKMUTR71dOI8Hg68kqJJ46nAQcDpDY7XifNYxmTjyhqvpEuAXcDqsaJJxpElPklPBi4BPlhr8SRj6ejvfCYmhY4+I1rSfhQJYXVEfD0VPyTpmLT8GGD7BLHmfA8vAM6UtBUYomhC+jRwmKSxhzJVH68SS1p+KLAjc4wjwEhE3JHmr6NIEt10Hl8G/CIiHo6IPwBfB55Pd53Haq06dyNpuuXxpo7YM4BzI7WrNBHfI9T/HUzFCRT/ANydPjvHAndJ+pMmYsx2DktpVztVt7wo/su8n+IXONYB9cw2HVvAl4FPjyv/JHt38v1dmn4Fe3dQ3ZnKj6BoUz88vX4BHJEh3kH2dDR/lb07596Wpt/O3h2ka9L0M9m7A/B+WtvR/D3gpDR9aTqHXXMegVOATcCT03G/BLyjW84jT+xTaNm5o3gO+/PY00n68hbEtwT4CXD0uPVqnh8afM7r/Q6mGuO4ZVvZ06fQkXPY9N9Guw7UTS+KqwHuo7g64ZI2HveFFNXAHwP/ll4vp2jnvAX4Wfo59och4MoU50ZgoGpfbwC2pNf5meIdZE9SOJ7iiogt6UN1QCp/UprfkpYfX7X9JSn2zbT46gngT4EfpnP5jfSh6qrzCHwY+ClwD/CP6Yur4+cRuJain+MPFP+VXtDKcwcMpPf8c+AKxl0Q0GR8Wyja38c+N1dNdH6o8zmv9zuYaozjlm9lT1Jo+zmcysvDXJiZWcVM7FMwM7M6nBTMzKzCScHMzCqcFMzMrMJJwczMKpwUrO3SCJIrq+bfK+nSFu37Gklnt2JfExznNSpGZ70197E6TdK70x27NgM4KVgn/B44a2xo4W4hadYkVr+A4qanxbni6SLvprgJz2YAJwXrhF0Uz6n9q/ELxv+nL2ln+jko6TZJayTdJ+njks6VdGcad/6Eqt28TNL30npnpO1npTH516cx7d9ctd9bJX2F4sai8fEsS/u/R9InUtkHKW5EvErSJ8etv4+kz6p4jsJaSevG3k8aI/82SRsk3Vw1rMSwpE+k93KfpP/ZKOYaMf5lWn63pH9MZU+XdEsqv0XS00qc32HteUbF6vQcgHdSjN10azpPs9I+7knn5Qm/Q+tx7bpLzi+/xl7ATorhr7dSjPHzXuDStOwa4OzqddPPQeA/KYYVPgDYBnw4LXsXaeiQtP2/UPzDcyLF3aZPApYDH0jrHEBxN/TctN9Hgbk14nwqxSinR1MMm/Ad4FVp2TBVd6ZWbXM2sC4d/0+A36Sy/YAfkIZpAP4cuLpqXyvT9MuBb6fpmjGPO94zKe7kHbt7duxO5BuB16fpNwDfKHF+RynG2dkHuB14YVq2tWr/C4FvVW1/WKf/nvxq7cs1BeuIKEaH/TLwzklstj4iHoyI31Pc/v+vqXwjxTg0Y9ZExO6I+BnF+Df/jeIhLX8p6d8ohis/kiJpQDEWzS9qHO85wHAUg9qNjcz5oglifCHw1XT8XwNjfQ4nAScD30oxfIC9Bz0bGxxxQ9V7aRTzmJcA10XEIwARsSOVL6J4SBIUQ2y8cIK4oTgPIxGxm2Ioif4a69wPHC/pcklLgEaj/FoP2nfiVcyy+TRwF/DFqrJdpGbN9AjC6kcl/r5qenfV/G72/lseP3bL2HDE74iIm6sXSBqkqCnU0swjEOttI2BTRCyqs3zsvfyRPe+lZsw19ltmrJqxdcqe3+o49uwk4jeSnkXxwKC3A6+lqInYNOGagnVM+q92DUWn7ZitFE0UUDyPYL8mdv2a1LZ/AsXgZ5spnsj11jR0OZKeoeLBPI3cAbxY0lGpE3oZcNsE23wf+N/p+H0UzTKkGI6WtCgdfz9Jz5xgX2VivgV4raQj0zpHpPIfUIy2CnBuiguaO7+/o3h8LOnigH0i4mvA31AMWW7TiGsK1mkrgQur5v8BuF7SnRRfePX+i29kM8WXdx/wloj4L0mfp2gOuSv9h/ww8KpGO4mIByVdTNEEJGBdRFw/wbG/BryUYoTL+ygSy2hEPJ46eC+TdCjFZ+/TFMNr1zNhzBGxSdJHgdsk/RH4EcWzoN8JXC3pfWm789MmzZzfVcA3JT1IcSXSFyWN/UN5cYntrYd4lFSzFpM0OyJ2pv/e7wRekPoXzLqeawpmrbdW0mEU7fUfcUKwXuKagpmZVbij2czMKpwUzMyswknBzMwqnBTMzKzCScHMzCr+P/C6dlpxSBz2AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "threshold = 0 # this filters the data by gene count\n",
    "bcv = [x for b,x in barcode_hist.items() if x > 0] \n",
    "_ = plt.hist(bcv,bins=100, log=True)\n",
    "plt.rcParams[\"figure.figsize\"] = [9,6]\n",
    "plt.xlabel(\"Number of gene counts\")\n",
    "plt.ylabel(\"Number of barcodes\")\n",
    "plt.grid(True)\n",
    "print(len(bcv))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export as .mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "outfile = './out_gmp_merged/matrix.mtx'\n",
    "\n",
    "gene_to_id = dict((g,i+1) for i,g in enumerate(genes))\n",
    "barcodes_to_use = [b for b,x in barcode_hist.items() if x > 500]\n",
    "\n",
    "num_entries = 0\n",
    "for barcode in barcodes_to_use:\n",
    "    num_entries += len([x for x in cell_gene[barcode].values() if round(x)>0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "72017"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(barcodes_to_use)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(outfile, 'w') as of:\n",
    "    of.write('%%MatrixMarket matrix coordinate real general\\n%\\n')\n",
    "    #number of genes\n",
    "    of.write(\"%d %d %d\\n\"%(len(genes), len(barcodes_to_use), num_entries))\n",
    "    bcid = 0\n",
    "    for barcode in barcodes_to_use:\n",
    "        bcid += 1\n",
    "        cg = cell_gene[barcode]\n",
    "        gl = [(gene_to_id[g],round(cg[g])) for g in cg if round(cg[g]) > 0]\n",
    "        gl.sort()\n",
    "        for x in gl:\n",
    "            of.write(\"%d %d %d\\n\"%(x[0],bcid,x[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = {}\n",
    "with open('./mouse_transcript_to_gene.tsv') as f:\n",
    "    f.readline()\n",
    "    for line in f:\n",
    "        t,g,gn = line.split()\n",
    "        gene_names[g] = gn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "id_to_genes = dict((i,g) for (g,i) in gene_to_id.items())\n",
    "gl = []\n",
    "for i in range(1,len(genes)+1):\n",
    "    g = id_to_genes[i]\n",
    "    gid = g[:g.find('.')]\n",
    "    if g in gene_names:\n",
    "        gn = gene_names[g]\n",
    "    else:\n",
    "        gn = ''\n",
    "    gl.append((g,gn))\n",
    "\n",
    "with open('./out_gmp_merged/genes_t.tsv','w') as of:\n",
    "    for g,gn in gl:\n",
    "        of.write(\"%s\\t%s\\n\"%(g,gn))\n",
    "        \n",
    "with open('./out_gmp_merged/barcodes.tsv','w') as of:\n",
    "    of.write('\\n'.join(x + '-1' for x in barcodes_to_use))\n",
    "    of.write('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSMUSG00000080735.1\n",
      "\n",
      "ENSMUSG00000109789.1\n",
      "\n",
      "ENSMUSG00000048281.9\n",
      "\n",
      "ENSMUSG00000099108.1\n",
      "\n",
      "ENSMUSG00000022253.15\n",
      "\n",
      "ENSMUSG00000097340.1\n",
      "\n",
      "ENSMUSG00000110118.1\n",
      "\n",
      "ENSMUSG00000058172.1\n",
      "\n",
      "ENSMUSG00000035158.15\n",
      "\n"
     ]
    }
   ],
   "source": [
    "id_to_genes = dict((i,g) for (g,i) in gene_to_id.items())\n",
    "gl = []\n",
    "for i in range(1,10):\n",
    "    g = id_to_genes[i]\n",
    "    gid = g[:g.find('.')]\n",
    "    print(g)\n",
    "    if g in gene_names:\n",
    "        gn = gene_names[g]\n",
    "    else:\n",
    "        gn = ''\n",
    "    print(gn)\n",
    "    gl.append((g,gn))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
